zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+
+            let mut outcome_info =
+                <Outcomes<T>>::get(market_id, &outcome).ok_or(Error::<T>::OutcomeDoesNotExist)?;
+            let outcome_owners_len = outcome_info.owners.len() as u32;
+
+            // The `outcome_sum` never decreases (only increases) to allow
+            // caching the outcome with the highest `outcome_sum`.
+            // If the `outcome_sum` decreases, it would lead to more storage,
+            // because the winning outcome could have a smaller `outcome_sum`
+            // than the second highest `outcome_sum`.
+            let add_to_outcome_sum = |a| {
+                outcome_info.outcome_sum = outcome_info.outcome_sum.saturating_add(a);
+                <Winners<T>>::mutate(market_id, |highest| {
+                    *highest = Some(highest.clone().map_or(
+                        WinnerInfo::new(outcome.clone(), outcome_info.outcome_sum),
+                        |prev_winner_info| {
+                            if outcome_info.outcome_sum >= prev_winner_info.vote_sum {
+                                WinnerInfo::new(outcome.clone(), outcome_info.outcome_sum)
+                            } else {
+                                prev_winner_info
+                            }
+                        },
+                    ));
+                });
+                <Outcomes<T>>::insert(market_id, &outcome, outcome_info);
+            };
+
+            let vote_lock_counter = <Locks<T>>::try_mutate(
+                &sender,
+                |LockInfo(ref mut lock_info)| -> Result<u32, DispatchError> {
+                    match lock_info.binary_search_by_key(&market_id, |i| i.0) {
+                        Ok(i) => {
+                            let prev_highest_amount: BalanceOf<T> = lock_info[i].1;
+                            // concious decision here to not throw an error when the new amount is not higher
+                            // because the user would have to pay worst case fees then
+                            if amount > prev_highest_amount {
+                                let diff = amount.saturating_sub(prev_highest_amount);
+                                add_to_outcome_sum(diff);
+                                lock_info[i].1 = amount;
+                            }
+                        }
+                        Err(i) => {
+                            ensure!(
+                                lock_info.len() as u32 <= T::MaxGlobalDisputeVotes::get(),
+                                Error::<T>::MaxVotesReached
+                            );
+                            add_to_outcome_sum(amount);
+                            lock_info.insert(i, (market_id, amount));
+                        }
+                    }
+                    Ok(lock_info.len() as u32)
+                },
+            )?;
+
+            T::Currency::extend_lock(
+                T::VoteLockIdentifier::get(),
+                &sender,
+                amount,
```

Comment:
On another note, since the lock currently uses only one identifier, we have to lock the sum of all lock amounts, and not just the highest amount. If we have one lock, for example 50 ZTG and create another lock by voting, for example 20 ZTG, the total lock amount will be 20 ZTG as of the current code instead of 70. So in case `amount > prev_highest_amount`, we have to get the previous amount in the lock for `VoteLockIdentifier` and add the `prev_highest_amount - amount`.